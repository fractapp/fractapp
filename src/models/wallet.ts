import { ImageSourcePropType } from "react-native";

export enum Currency {
    Polkadot = 0,
    Kusama
}

export function getLogo(currency: Currency) {
    let logo: ImageSourcePropType
    switch (currency) {
        case Currency.Polkadot:
            logo = require("assets/img/dot.png")
            break
        case Currency.Kusama:
            logo = require("assets/img/kusama.png")
            break
        default:
            throw ("invalid currency")
    }
    return logo
}
export function getSymbol(currency: Currency) {
    let symbol = ""
    switch (currency) {
        case Currency.Polkadot:
            symbol = "DOT"
            break
        case Currency.Kusama:
            symbol = "KSM"
            break
        default:
            throw ("invalid currency")
    }
    return symbol
}

export class Wallet {
    name: string;
    address: string;
    balance: number;
    currency: Currency;
    usdValue: number;

    constructor(name: string, address: string, currency: Currency, balance: number, price: number) {
        this.name = name;
        this.address = address;
        this.balance = balance;
        this.currency = currency;

        this.usdValue = +(balance * price).toFixed(2)
    }
}
