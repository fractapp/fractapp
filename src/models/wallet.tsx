import { ImageSourcePropType } from "react-native";

export enum Currency {
    Polkadot = 0,
    Kusama
}

export class Wallet {
    name: string;
    address: string;
    balance: number;
    currency: Currency;

    symbol: string;
    usdValue: number;
    logo: ImageSourcePropType;

    constructor(name: string, address: string, currency: Currency, balance: number) {
        this.name = name;
        this.address = address;
        this.balance = balance;
        this.currency = currency;
        
        //TODO get state
        let rate = 10
        this.usdValue = balance * rate
        switch (currency) {
            case Currency.Polkadot:
                this.logo = require("assets/img/dot.png")
                this.symbol = "DOT"
                break
            case Currency.Kusama:
                this.logo = require("assets/img/kusama.png")
                this.symbol = "KSM"
                break
        }

    }
}
