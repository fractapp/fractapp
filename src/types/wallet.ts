import {ImageSourcePropType} from 'react-native';

/**
 * @category Models
 */
export enum Currency {
  Polkadot = 0,
  Kusama,
}

/**
 * get logo for wallet
 * @category Models
 */
export function getLogo(currency: Currency) {
  let logo: ImageSourcePropType;
  switch (currency) {
    case Currency.Polkadot:
      logo = require('assets/img/dot.png');
      break;
    case Currency.Kusama:
      logo = require('assets/img/kusama.png');
      break;
    default:
      throw new Error('invalid currency');
  }
  return logo;
}

/**
 * check logo without border
 * @category Models
 */
export function withoutBorder(currency: Currency): boolean {
  return currency === Currency.Polkadot || currency === Currency.Kusama;
}

/**
 * get symbol for wallet
 * @category Models
 */
export function getSymbol(currency: Currency) {
  let symbol = '';
  switch (currency) {
    case Currency.Polkadot:
      symbol = 'DOT';
      break;
    case Currency.Kusama:
      symbol = 'KSM';
      break;
    default:
      throw new Error('invalid currency');
  }
  return symbol;
}

/**
 * @category Models
 */
export class Wallet {
  name: string;
  address: string;
  balance: number;
  planks: string;
  currency: Currency;
  usdValue: number;

  constructor(
    name: string,
    address: string,
    currency: Currency,
    balance: number,
    planks: string,
    price: number,
  ) {
    this.name = name;
    this.address = address;
    this.balance = balance;
    this.planks = planks;
    this.currency = currency;

    this.usdValue = +(balance * price).toFixed(2);
  }
}
