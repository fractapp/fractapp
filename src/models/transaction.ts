import {Currency} from './wallet';

/**
 * Transaction type
 * @category Models
 */
export enum TxType {
  None = 0,
  Sent,
  Received,
}

export enum TxStatus {
  Pending = 0,
  Success,
  Fail,
}

/**
 * @category Models
 */
export class Transaction {
  id: string;
  member: string;
  currency: Currency;
  value: number;
  txType: TxType;
  timestamp: number;
  fee: number;
  usdValue: number;
  usdFee: number;
  status: TxStatus;

  constructor(
    id: string,
    member: string,
    currency: Currency,
    txType: TxType,
    timestamp: number,
    value: number,
    usdValue: number,
    fee: number,
    usdFee: number,
    status: TxStatus,
  ) {
    this.id = id;
    this.member = member;
    this.currency = currency;
    this.value = value;
    this.txType = txType;
    this.timestamp = timestamp;
    this.fee = fee;
    this.usdValue = usdValue;
    this.usdFee = usdFee;
    this.status = status;
  }
}
