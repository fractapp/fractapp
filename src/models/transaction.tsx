import { ImageSourcePropType } from "react-native";
import { Currency } from "./wallet";

export enum TxType {
    None = 0,
    Sent,
    Received
}

export class Transaction {
    id: string;
    member: string;
    currency: Currency;
    value: number;
    txType: TxType;
    date: Date;

    usdValue: number;
    symbol: string;
    logo: ImageSourcePropType;

    constructor(id: string, member: string, currency: Currency, txType: TxType, date: Date, value: number) {
        this.id = id;
        this.member = member;
        this.currency = currency;
        this.value = value;
        this.txType = txType;
        this.date = date;
        //TODO get state
        let rate = 10
        this.usdValue = value * rate
        switch (currency) {
            case Currency.Polkadot:
                this.logo = require("assets/img/dot.png")
                this.symbol = "DOT"
                break
            case Currency.Kusama:
                this.logo = require("assets/img/kusama.png")
                this.symbol = "KSM"
                break
        }

    }
}
