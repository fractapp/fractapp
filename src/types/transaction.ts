import {Currency} from './/wallet';

/**
 * Transaction type
 * @category Models
 */
export enum TxType {
  None = 0,
  Sent,
  Received,
}

/**
 * Transaction action type
 * @category Models
 */
export enum TxAction {
  Transfer,
  StakingReward,
  StakingCreateWithdrawalRequest,
  StakingWithdrawn,
  StakingOpenDeposit,
  StakingAddAmount,
  ConfirmWithdrawal,
  UpdateNomination
}

export enum TxStatus {
  Pending = 0,
  Success,
  Fail,
}

/**
 * @category Models
 */
export type Transaction = {
  id: string;
  hash: string;
  userId: string | null;

  address: string;
  currency: Currency;
  action: TxAction,
  txType: TxType;
  timestamp: number;

  value: number;
  planckValue: string;
  usdValue: number;
  fullValue: string;

  fee: number;
  planckFee: string;
  usdFee: number;

  status: TxStatus;
};


/**
 * @category Models
 */
export type ApiTransaction = {
  id: string;
  hash: string;
  currency: Currency;

  from: string;
  userFrom: string;

  action: TxAction,

  to: string;
  userTo: string;

  value: string;
  fee: string;
  price: number

  timestamp: number;
  status: TxStatus;
};

