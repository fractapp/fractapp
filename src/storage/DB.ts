import {Account} from 'models/account';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Keychain from 'react-native-keychain';
import {base64Encode, randomAsU8a} from '@polkadot/util-crypto';
import PasscodeUtil from 'utils/passcode';
import {Keyring} from '@polkadot/keyring';
import {u8aToHex} from '@polkadot/util';
import {Currency, getSymbol} from 'models/wallet';
import {Transaction} from 'models/transaction';
import {ChatInfo} from 'models/chatInfo';
import {AuthInfo} from 'models/authInfo';
import {MyProfile} from 'models/myProfile';
import BN from 'bn.js';
import {UserProfile} from 'models/profile';
import {Use} from 'react-native-svg';
/**
 * @namespace
 * @category storage
 */

namespace DB {
  const secureOption = {
    service: 'com.fractapp',
  };
  const biometryOption = {
    service: 'com.fractapp',
    accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_CURRENT_SET,
  };

  const AsyncStorageKeys = {
    price: (currency: Currency) => `price_${currency}`,
    profile: 'profile',
    authInfo: 'auth_info',
    notificationCount: 'notification_count',
    accounts: 'accounts',
    contacts: 'contacts',
    users: 'users',
    accountInfo: (address: string) => `account_${address}`,
    transactions: (currency: Currency) => `transactions_${getSymbol(currency)}`,
    pendingTransactions: (currency: Currency) =>
      `pending_transactions_${getSymbol(currency)}`,
    chatByAddress: (address: string) => `chat_${address}`,
    chatsInfo: 'chats_info',
  };
  const SecureStorageKeys = {
    firebaseToken: 'firebase_token',
    authJWT: 'auth_jwt',
    seed: 'seed',
    salt: 'salt',
    passcodeHash: 'passcode_hash',
  };
  const PasscodeStorageKeys = {
    passcode: 'passcode',
  };

  export async function setSecureItem(key: string, value: string) {
    const result = await Keychain.setInternetCredentials(
      key,
      key,
      value,
      secureOption,
    );
    if (result == false) {
      throw `invalid set ${key}`;
    }
  }
  export async function getSecureItem(key: string): Promise<string | null> {
    const result = await Keychain.getInternetCredentials(key, secureOption);
    if (result === false) {
      return null;
    }

    return result.password;
  }

  export async function setAuthInfo(auth: AuthInfo) {
    await AsyncStorage.setItem(AsyncStorageKeys.authInfo, JSON.stringify(auth));
  }
  export async function getAuthInfo(): Promise<AuthInfo | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.authInfo);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setPrice(currency: Currency, value: number) {
    await AsyncStorage.setItem(AsyncStorageKeys.price(currency), String(value));
  }
  export async function getPrice(currency: Currency): Promise<number> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.price(currency));
    if (result == null) {
      return 0;
    }
    return Number(result);
  }

  export async function setProfile(profile: MyProfile) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.profile,
      JSON.stringify(profile),
    );
  }
  export async function getProfile(): Promise<MyProfile | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.profile);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setNotificationCount(count: number) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.notificationCount,
      String(count),
    );
  }
  export async function getNotificationCount(): Promise<number> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.notificationCount,
    );
    return Number(result);
  }

  export async function enablePasscode(passcode: string, isBiometry: boolean) {
    if (isBiometry) {
      const result = await Keychain.setInternetCredentials(
        PasscodeStorageKeys.passcode,
        PasscodeStorageKeys.passcode,
        passcode,
        biometryOption,
      );
      if (result === false) {
        throw 'invalid set passcode';
      }
    }

    const salt = base64Encode(randomAsU8a(32));
    const hash = PasscodeUtil.hash(passcode, salt);
    let authInfo = await getAuthInfo();
    if (authInfo == null) {
      authInfo = new AuthInfo(false, false, false, false);
    }

    try {
      await setSecureItem(SecureStorageKeys.salt, salt);
      await setSecureItem(SecureStorageKeys.passcodeHash, hash);
      authInfo.isPasscode = true;
      authInfo.isBiometry = isBiometry;
    } catch (e) {
      await disablePasscode();
    }

    await setAuthInfo(authInfo);
  }
  export async function disablePasscode() {
    let authInfo = await getAuthInfo();
    if (authInfo == null) {
      authInfo = new AuthInfo(false, false, false, false);
    }
    authInfo.isPasscode = false;
    authInfo.isBiometry = false;

    await setAuthInfo(authInfo);

    await Keychain.resetInternetCredentials(
      SecureStorageKeys.salt,
      secureOption,
    );
    await Keychain.resetInternetCredentials(
      SecureStorageKeys.passcodeHash,
      secureOption,
    );
  }

  export async function getSalt(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.salt);
  }
  export async function getPasscodeHash(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.passcodeHash);
  }
  export async function getPasscode(): Promise<string> {
    const result = await Keychain.getInternetCredentials(
      PasscodeStorageKeys.passcode,
      biometryOption,
    );
    if (result === false) {
      throw 'value by passcode key not found';
    }

    return result.password;
  }

  export async function createAccounts(seed: string) {
    let polkadotWallet = new Keyring({
      type: 'sr25519',
      ss58Format: 0,
    }).addFromUri(seed);
    let kusamaWallet = new Keyring({type: 'sr25519', ss58Format: 2}).addFromUri(
      seed,
    );
    let accountsInfo = new Array<Account>(
      new Account(
        'Polkadot wallet',
        polkadotWallet.address,
        u8aToHex(polkadotWallet.publicKey),
        Currency.Polkadot,
        0,
        new BN(0).toString(10),
      ),
      new Account(
        'Kusama wallet',
        kusamaWallet.address,
        u8aToHex(kusamaWallet.publicKey),
        Currency.Kusama,
        0,
        new BN(0).toString(10),
      ),
    );
    let accounts = new Array<string>();
    for (let i = 0; i < accountsInfo.length; i++) {
      const account = accountsInfo[i];
      accounts.push(account.address);
      await setAccountInfo(account);
    }
    await setAccounts(accounts);
    await setSecureItem(SecureStorageKeys.seed, seed);
  }
  export async function getSeed(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.seed);
  }

  export async function setAccounts(accounts: Array<string>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.accounts,
      JSON.stringify(accounts),
    );
  }
  export async function getAccounts(): Promise<Array<string> | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.accounts);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setChatsInfo(chats: Map<string, ChatInfo>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.chatsInfo,
      JSON.stringify([...chats]),
    );
  }
  export async function getChatsInfo(): Promise<Map<string, ChatInfo>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.chatsInfo);

    if (result == null) {
      return new Map<string, ChatInfo>();
    }
    return new Map<string, ChatInfo>(JSON.parse(result));
  }

  export async function setChat(address: string, chats: Map<string, Currency>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.chatByAddress(address),
      JSON.stringify([...chats]),
    );
  }
  export async function getChat(
    address: string,
  ): Promise<Map<string, boolean>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.chatByAddress(address),
    );

    if (result == null) {
      return new Map<string, boolean>();
    }
    return new Map<string, boolean>(JSON.parse(result));
  }

  export async function setTx(currency: Currency, tx: Transaction) {
    const txs = await getTxs(currency);

    txs.set(tx.id, tx);
    await AsyncStorage.setItem(
      AsyncStorageKeys.transactions(currency),
      JSON.stringify([...txs]),
    );
  }
  export async function getTxs(
    currency: Currency,
  ): Promise<Map<string, Transaction>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.transactions(currency),
    );

    if (result == null) {
      return new Map<string, Transaction>();
    }
    return new Map<string, Transaction>(JSON.parse(result));
  }

  export async function setContacts(contacts: Array<UserProfile>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.contacts,
      JSON.stringify(contacts),
    );
  }

  export async function getContacts(): Promise<Array<UserProfile>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.contacts);

    if (result == null) {
      return [];
    }

    return JSON.parse(result);
  }

  export async function setUsers(users: Map<string, UserProfile>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.users,
      JSON.stringify([...users]),
    );
  }

  export async function getUsers(): Promise<Map<string, UserProfile>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.users);

    if (result == null) {
      return new Map<string, UserProfile>();
    }
    return new Map<string, UserProfile>(JSON.parse(result));
  }

  export async function setPendingTxs(currency: Currency, txs: Array<string>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.pendingTransactions(currency),
      JSON.stringify(txs),
    );
  }

  export async function getPendingTxs(
    currency: Currency,
  ): Promise<Array<string>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.pendingTransactions(currency),
    );

    if (result == null) {
      return [];
    }

    return JSON.parse(result);
  }

  export async function setAccountInfo(account: Account) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.accountInfo(account.address),
      JSON.stringify(account),
    );
  }
  export async function getAccountInfo(
    address: string,
  ): Promise<Account | null> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.accountInfo(address),
    );
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setFirebaseToken(token: string) {
    await setSecureItem(SecureStorageKeys.firebaseToken, token);
  }
  export async function getFirebaseToken(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.firebaseToken);
  }

  export async function setJWT(token: string) {
    await setSecureItem(SecureStorageKeys.authJWT, token);
  }
  export async function getJWT(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.authJWT);
  }
}

export default DB;
