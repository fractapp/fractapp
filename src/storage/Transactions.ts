import {Currency} from 'types/wallet';
import {createContext, Dispatch} from 'react';
import {Transaction} from 'types/transaction';
import DB from 'storage/DB';

/**
 * @namespace
 * @category Context storage
 */
namespace TransactionsStore {
  export enum Action {
    SET,
    SET_TX,
    ADD_PENDING_TXS,
    REMOVE_PENDING_TXS,
  }

  export type State = {
    transactions: Map<Currency, Map<string, Transaction>>;
    pendingTransactions: Map<Currency, Array<string>>;
    isInitialized: boolean;
  };

  export const initialState = (): State => ({
    transactions: new Map<Currency, Map<string, Transaction>>(),
    pendingTransactions: new Map<Currency, Array<string>>(),
    isInitialized: false,
  });

  export type ContextType = {
    state: State;
    dispatch: Dispatch<any>;
  };

  export const Context = createContext<ContextType>({
    state: initialState(),
    dispatch: () => null,
  });

  export function reducer(prevState: State, action: any): State {
    let copy: State = Object.assign({}, prevState);

    switch (action.type) {
      case Action.SET:
        copy.transactions = action.txs;
        copy.pendingTransactions = action.pendingTxs;
        copy.isInitialized = true;
        return copy;
      case Action.SET_TX:
        if (!copy.transactions.has(action.currency)) {
          copy.transactions.set(
            action.currency,
            new Map<string, Transaction>(),
          );
        }

        copy.transactions.get(action.currency)?.set(action.tx.id, action.tx);

        DB.setTx(action.currency, action.tx);
        return copy;
      case Action.ADD_PENDING_TXS:
        if (!copy.pendingTransactions.has(action.currency)) {
          copy.pendingTransactions.set(action.currency, []);
        }

        copy.pendingTransactions.get(action.currency)?.push(action.hash);
        DB.setPendingTxs(
          action.currency,
          copy.pendingTransactions.get(action.currency)!,
        );
        return copy;
      case Action.REMOVE_PENDING_TXS:
        copy.pendingTransactions.get(action.currency)?.splice(action.index, 1);

        DB.setPendingTxs(
          action.currency,
          copy.pendingTransactions.get(action.currency)!,
        );
        return copy;
      default:
        return prevState;
    }
  }

  export const set = (
    txs: Map<Currency, Map<string, Transaction>>,
    pendingTxs: Map<Currency, Array<string>>,
  ) => ({
    type: Action.SET,
    txs: txs,
    pendingTxs: pendingTxs,
  });
  export const setTx = (currency: Currency, tx: Transaction) => ({
    type: Action.SET_TX,
    tx: tx,
    currency: currency,
  });
  export const addPendingTx = (currency: Currency, hash: string) => ({
    type: Action.ADD_PENDING_TXS,
    hash: hash,
    currency: currency,
  });
  export const removePendingTx = (currency: Currency, index: number) => ({
    type: Action.REMOVE_PENDING_TXS,
    index: index,
    currency: currency,
  });
}
export default TransactionsStore;
