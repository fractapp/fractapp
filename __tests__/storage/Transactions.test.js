import {Currency} from 'types/wallet';
import DB from 'storage/DB';
import Transactions from 'storage/Transactions';
import {TxStatus, TxType} from 'types/transaction';

jest.mock('storage/DB', () => ({
  setPendingTxs: jest.fn(),
  setTx: jest.fn(),
}));

const initState = () => ({
  transactions: new Map(),
  pendingTransactions: new Map(),
  isInitialized: false,
});

it('Test set', async () => {
  const txs = new Map([
    [
      Currency.Polkadot,
      new Map([
        [
          'txIdOne',
          {
            id: 'txIdOne',
            userId: 'userId',
            address: 'address#1',
            currency: Currency.Polkadot,
            txType: TxType.None,
            timestamp: new Date('12-12-2020').getTime(),
            value: 10,
            usdValue: 10,
            fee: 10,
            usdFee: 10,
            status: TxStatus.Success,
          },
        ],
      ]),
    ],
  ]);
  const pendingTxs = new Map([[Currency.Polkadot, ['txOne', 'txTwo']]]);
  expect(Transactions.set(txs, pendingTxs)).toMatchSnapshot();
});
it('Test setTx', async () => {
  const tx = {
    id: 'txIdOne',
    userId: 'userId',
    address: 'address#1',
    currency: Currency.Kusama,
    txType: TxType.Received,
    timestamp: new Date('12-12-2020').getTime(),
    value: 10,
    usdValue: 10,
    fee: 10,
    usdFee: 10,
    status: TxStatus.Success,
  };
  expect(Transactions.setTx(Currency.Kusama, tx)).toMatchSnapshot();
});
it('Test addPendingTx', async () => {
  expect(
    Transactions.addPendingTx(Currency.Kusama, 'txHash'),
  ).toMatchSnapshot();
});
it('Test removePendingTx', async () => {
  expect(Transactions.removePendingTx(Currency.Kusama, 10)).toMatchSnapshot();
});

it('Test reducer set', async () => {
  const txs = new Map([
    [
      Currency.Polkadot,
      new Map([
        [
          'txIdOne',
          {
            id: 'txIdOne',
            userId: 'userId',
            address: 'address#1',
            currency: Currency.Polkadot,
            txType: TxType.None,
            timestamp: new Date('12-12-2020').getTime(),
            value: 10,
            usdValue: 10,
            fee: 10,
            usdFee: 10,
            status: TxStatus.Success,
          },
        ],
      ]),
    ],
  ]);
  const pendingTxs = new Map([[Currency.Polkadot, ['txOne', 'txTwo']]]);
  expect(
    Transactions.reducer(initState(), Transactions.set(txs, pendingTxs)),
  ).toMatchSnapshot();
});
it('Test reducer setTx', async () => {
  const tx = {
    id: 'txIdOne',
    userId: 'userId',
    address: 'address#1',
    currency: Currency.Kusama,
    txType: TxType.Received,
    timestamp: new Date('12-12-2020').getTime(),
    value: 10,
    usdValue: 10,
    fee: 10,
    usdFee: 10,
    status: TxStatus.Success,
  };
  expect(
    Transactions.reducer(initState(), Transactions.setTx(tx.currency, tx)),
  ).toMatchSnapshot();
  expect(DB.setTx).toBeCalledWith(tx.currency, tx);
});
it('Test reducer addPendingTx', async () => {
  expect(
    Transactions.reducer(
      initState(),
      Transactions.addPendingTx(Currency.Kusama, 'txHash'),
    ),
  ).toMatchSnapshot();
  expect(DB.setPendingTxs).toBeCalledWith(Currency.Kusama, ['txHash']);
});
it('Test reducer removePendingTx', async () => {
  let state = initState();
  state = Transactions.reducer(
    state,
    Transactions.addPendingTx(Currency.Kusama, 'txHashOne'),
  );
  expect(state).toMatchSnapshot();
  expect(DB.setPendingTxs).toBeCalledWith(Currency.Kusama, ['txHashOne']);

  state = Transactions.reducer(
    state,
    Transactions.addPendingTx(Currency.Kusama, 'txHashTwo'),
  );
  expect(state).toMatchSnapshot();
  expect(DB.setPendingTxs).toBeCalledWith(Currency.Kusama, [
    'txHashOne',
    'txHashTwo',
  ]);

  expect(
    Transactions.reducer(
      state,
      Transactions.removePendingTx(Currency.Kusama, 1),
    ),
  ).toMatchSnapshot();
  expect(DB.setPendingTxs).toBeCalledWith(Currency.Kusama, ['txHashOne']);
});

it('Test default', async () => {
  expect(
    Transactions.reducer(Transactions.initialState(), {
      type: 9999,
    }),
  ).toStrictEqual(initState());
});
