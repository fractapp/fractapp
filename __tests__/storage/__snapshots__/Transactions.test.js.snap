// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test addPendingTx 1`] = `
Object {
  "currency": 1,
  "hash": "txHash",
  "type": 2,
}
`;

exports[`Test reducer addPendingTx 1`] = `
Object {
  "isInitialized": false,
  "pendingTransactions": Map {
    1 => Array [
      "txHash",
    ],
  },
  "transactions": Map {},
}
`;

exports[`Test reducer removePendingTx 1`] = `
Object {
  "isInitialized": false,
  "pendingTransactions": Map {
    1 => Array [
      "txHashOne",
    ],
  },
  "transactions": Map {},
}
`;

exports[`Test reducer removePendingTx 2`] = `
Object {
  "isInitialized": false,
  "pendingTransactions": Map {
    1 => Array [
      "txHashOne",
      "txHashTwo",
    ],
  },
  "transactions": Map {},
}
`;

exports[`Test reducer removePendingTx 3`] = `
Object {
  "isInitialized": false,
  "pendingTransactions": Map {
    1 => Array [
      "txHashOne",
    ],
  },
  "transactions": Map {},
}
`;

exports[`Test reducer set 1`] = `
Object {
  "isInitialized": true,
  "pendingTransactions": Map {
    0 => Array [
      "txOne",
      "txTwo",
    ],
  },
  "transactions": Map {
    0 => Map {
      "txIdOne" => Object {
        "address": "address#1",
        "currency": 0,
        "fee": 10,
        "id": "txIdOne",
        "status": 1,
        "timestamp": 1607720400000,
        "txType": 0,
        "usdFee": 10,
        "usdValue": 10,
        "userId": "userId",
        "value": 10,
      },
    },
  },
}
`;

exports[`Test reducer setTx 1`] = `
Object {
  "isInitialized": false,
  "pendingTransactions": Map {},
  "transactions": Map {
    1 => Map {
      "txIdOne" => Object {
        "address": "address#1",
        "currency": 1,
        "fee": 10,
        "id": "txIdOne",
        "status": 1,
        "timestamp": 1607720400000,
        "txType": 2,
        "usdFee": 10,
        "usdValue": 10,
        "userId": "userId",
        "value": 10,
      },
    },
  },
}
`;

exports[`Test removePendingTx 1`] = `
Object {
  "currency": 1,
  "index": 10,
  "type": 3,
}
`;

exports[`Test set 1`] = `
Object {
  "pendingTxs": Map {
    0 => Array [
      "txOne",
      "txTwo",
    ],
  },
  "txs": Map {
    0 => Map {
      "txIdOne" => Object {
        "address": "address#1",
        "currency": 0,
        "fee": 10,
        "id": "txIdOne",
        "status": 1,
        "timestamp": 1607720400000,
        "txType": 0,
        "usdFee": 10,
        "usdValue": 10,
        "userId": "userId",
        "value": 10,
      },
    },
  },
  "type": 0,
}
`;

exports[`Test setTx 1`] = `
Object {
  "currency": 1,
  "tx": Object {
    "address": "address#1",
    "currency": 1,
    "fee": 10,
    "id": "txIdOne",
    "status": 1,
    "timestamp": 1607720400000,
    "txType": 2,
    "usdFee": 10,
    "usdValue": 10,
    "userId": "userId",
    "value": 10,
  },
  "type": 1,
}
`;
